### PR TITLE
openbox: Use python3 instead of python2

### PR DESCRIPTION
**Pull request by SFrijters**
###### Motivation for this change

Openbox contains a single python script that is trivially convertible to python3 (print statements to print functions). With an automatic conversion in the patch phase we can drop the python2 dependency.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by doronbehar** [comment]
Could you show a diff please of the changes `2to3 does to the file? I'm thinking it might be more reliable to add a static patch that would do that. Arch Linux has such patches:

https://git.archlinux.org/svntogit/community.git/tree/trunk?h=packages/openbox

**Comment by SFrijters** [comment]
Sure, I'm also fine with a static patch. The autogenerated changes are pretty much the same as in Arch (comparing to https://git.archlinux.org/svntogit/community.git/tree/trunk/py3.patch?h=packages/openbox). They modify the first chunk of the python script slightly differently, and they also patch some autoconfig files, which we won't need to do because we rewrite the shebangs anyway. Just let me know what you prefer, I'll just be happy to get rid of the last python2 thing in my nix store.

```
32,34c32,34
<     print
<     print >>sys.stderr, "ERROR:", ME, "requires PyXDG to be installed"
<     print
---
>     print()
>     print("ERROR:", ME, "requires PyXDG to be installed", file=sys.stderr)
>     print()
55c55
<                 print "Invalid .desktop file: " + path
---
>                 print("Invalid .desktop file: " + path)
103c103
<             print "\t ", str
---
>             print("\t ", str)
105c105
<             print "\t*", str
---
>             print("\t*", str)
150c150
<             print "[*] " + self.de.getName()
---
>             print("[*] " + self.de.getName())
152c152
<             print "[ ] " + self.de.getName()
---
>             print("[ ] " + self.de.getName())
157c157
<         print
---
>         print()
169,191c169,191
<     print "Usage:", ME, "[OPTION]... [ENVIRONMENT]..."
<     print
<     print "This tool will run xdg autostart .desktop files"
<     print
<     print "OPTIONS"
<     print "  --list        Show a list of the files which would be run"
<     print "                Files which would be run are marked with an asterix"
<     print "                symbol [*].  For files which would not be run,"
<     print "                information is given for why they are excluded"
<     print "  --help        Show this help and exit"
<     print "  --version     Show version and copyright information"
<     print
<     print "ENVIRONMENT specifies a list of environments for which to run autostart"
<     print "applications.  If none are specified, only applications which do not "
<     print "limit themselves to certain environments will be run."
<     print
<     print "ENVIRONMENT can be one or more of:"
<     print "  GNOME         Gnome Desktop"
<     print "  KDE           KDE Desktop"
<     print "  ROX           ROX Desktop"
<     print "  XFCE          XFCE Desktop"
<     print "  Old           Legacy systems"
<     print
---
>     print("Usage:", ME, "[OPTION]... [ENVIRONMENT]...")
>     print()
>     print("This tool will run xdg autostart .desktop files")
>     print()
>     print("OPTIONS")
>     print("  --list        Show a list of the files which would be run")
>     print("                Files which would be run are marked with an asterix")
>     print("                symbol [*].  For files which would not be run,")
>     print("                information is given for why they are excluded")
>     print("  --help        Show this help and exit")
>     print("  --version     Show version and copyright information")
>     print()
>     print("ENVIRONMENT specifies a list of environments for which to run autostart")
>     print("applications.  If none are specified, only applications which do not ")
>     print("limit themselves to certain environments will be run.")
>     print()
>     print("ENVIRONMENT can be one or more of:")
>     print("  GNOME         Gnome Desktop")
>     print("  KDE           KDE Desktop")
>     print("  ROX           ROX Desktop")
>     print("  XFCE          XFCE Desktop")
>     print("  Old           Legacy systems")
>     print()
194,196c194,196
<     print ME, VERSION
<     print "Copyright (c) 2008        Dana Jansens"
<     print
---
>     print(ME, VERSION)
>     print("Copyright (c) 2008        Dana Jansens")
>     print()
```

**Comment by doronbehar** [comment]
> Just let me know what you prefer, I'll just be happy to get rid of the last python2 thing in my nix store.

I feel you body :) I also had some PRs motivated by getting rid of Python2. Not all of them have even reached the approval state. Also, note I'm only a member of `@NixOS/nixpkgs-maintainers` as well so the best I'll be able to help you is approve the PR.

My personal opinion, also judging from other PRs I saw that added patches similar to or based on patches from other distros (#84072 ), I think it'll be best to download that patch from Arch and put in `openbox/py3.patch` and add it as `./py3.patch`.

**Comment by SFrijters** [comment]
@FRidh Could you maybe weigh in on this and add it to the Python 2 deprecation project?

**Comment by veprbl** [comment]
Upstream PR https://github.com/Mikachu/openbox/pull/9

> , I think it'll be best to download that patch from Arch and put in openbox/py3.patch and add it as ./py3.patch.

Please use `fetchpatch`.

**Comment by SFrijters** [comment]
@veprbl I tried to implement your requested change but I ran into two problems

* The patch from the PR you linked doesn't apply to the tarball nixpkgs uses. I tried to switch over our `src` to pull from GitHub, but then we need an additional `./bootstrap` command, which doesn't seem to play nicely with the `autoreconfHook`?
* The PR patch includes a rename which doesn't work with fetchpatch (https://github.com/NixOS/nixpkgs/issues/32084). Indeed when I do a build with `-K` and look in the build dir, files are not correctly renamed.

**Comment by veprbl** [comment]
@SFrijters Does this not work?
```nix
fetchpatch {
  name = "openbox-py3.patch";
  url = "https://git.archlinux.org/svntogit/community.git/plain/trunk/py3.patch?h=packages/openbox&id=147b9734e4bbdf5558e3f7d11c6541e2c429203c";
  sha256 = "019q8yq334wfrhmbh58wsh3cfpbj48pf08s5nlp51nak7dkn0hgi";
}
```

**Comment by SFrijters** [comment]
@veprbl Yes, that patch does apply. But I'm still left with a `#!@PYTHON@` shebang in my installation, so the script doesn't work properly anymore.

The build log does say `Rewriting #!@PYTHON@ to #!/nix/store/2dcsn57cgaxs92ha5swihrab0g3l2h6g-python3-3.7.7`, but it doesn't actually happen.

```
$ /nix/store/py75wi3hislh1llgmillg074lr43z8as-openbox-3.6.1/libexec/openbox-xdg-autostart
bash: /nix/store/py75wi3hislh1llgmillg074lr43z8as-openbox-3.6.1/libexec/openbox-xdg-autostart: @PYTHON@: bad interpreter: No such file or directory
```

In this case we would also need to use autoconf again to make sure the file is configured correctly, but then we're still left with the problem that the renaming of the file `data/autostart/openbox-xdg-autostart` to `data/autostart/openbox-xdg-autostart.in` doesn't actually work.

**Comment by SFrijters** [comment]
Pushed a new attempt. It uses fetchpatch, but then also includes a workaround to rename the required files and an autoconf step so they are actually configured. And since the tarball src is already autoconfed I switched to the github source. At least the patch is now a PR on the actual src.

File: pkgs/applications/window-managers/openbox/default.nix
Comment (veprbl) [inline pkgs/applications/window-managers/openbox/default.nix:43]: Ok, I think renames work with `fetchurl`, let's switch to that as an exception.

File: pkgs/applications/window-managers/openbox/default.nix
Comment (veprbl) [inline pkgs/applications/window-managers/openbox/default.nix:38]: Stick with the release tarball, if possible.

File: pkgs/applications/window-managers/openbox/default.nix
Comment (veprbl) [inline pkgs/applications/window-managers/openbox/default.nix:13]: Have you tried `autoreconfHook` instead?

File: pkgs/applications/window-managers/openbox/default.nix
Comment (veprbl) [inline pkgs/applications/window-managers/openbox/default.nix:45]: This is not a stable url. It may change contents in the case of force push (unlikely in this case, but still).
Could fetch individual commits:

* https://github.com/Mikachu/openbox/commit/acfbbc4ea40932f183617bb7006700140fe5f61e.patch
* https://github.com/Mikachu/openbox/commit/bd0ca516f57a7ee6d43c663d09a275953ce56b4e.diff

or just the first patch from AUR

* https://git.archlinux.org/svntogit/community.git/plain/openbox/trunk/py3.patch?id=90cb57ef53d952bb6ab4c33a184f815bbe1791c0

File: pkgs/applications/window-managers/openbox/default.nix
Comment (SFrijters) [reply pkgs/applications/window-managers/openbox/default.nix]: Ah, this helps avoid a lot of the problems. Seems everything falls into place now.

File: pkgs/applications/window-managers/openbox/default.nix
Comment (SFrijters) [reply pkgs/applications/window-managers/openbox/default.nix]: Tried it before, but now it works. Not that I've added `python3` to nativeBuildInputs as well to allow the autoreconf to find the python interpreter to set.

File: pkgs/applications/window-managers/openbox/default.nix
Comment (SFrijters) [reply pkgs/applications/window-managers/openbox/default.nix]: Used the AUR one since that one now correctly applies to the tarball.

**Comment by SFrijters** [comment]
@veprbl Thanks for shepherding this; I hope I have done it correctly this time around, with minimal changes.

File: pkgs/applications/window-managers/openbox/default.nix
Comment (veprbl) [inline pkgs/applications/window-managers/openbox/default.nix:14]: I believe, in this case, `python3` belongs to `buildInputs`.

Comment (SFrijters) [reply pkgs/applications/window-managers/openbox/default.nix]: Hm, not sure I understand this, but fixed nonetheless.

**Review by veprbl — APPROVED** [review]
Looks good. Thank you!

**Pull request merged**
